File: src/main/java/gargoyle/ct/messages/impl/CTComponentLocaleProvider.java
package gargoyle.ct.messages.impl;

import gargoyle.ct.messages.LocaleProvider;
import gargoyle.ct.pref.CTPreferences.SUPPORTED_LOCALES;
import gargoyle.ct.prop.CTObservableProperty;
import gargoyle.ct.ui.prop.CTComponentLocaleProperty;
import org.jetbrains.annotations.NotNull;

import java.awt.*;
import java.util.Locale;

public final class CTComponentLocaleProvider implements LocaleProvider {
    private final CTObservableProperty<SUPPORTED_LOCALES> locale;

    public CTComponentLocaleProvider(@NotNull Component component) {
        this(new CTComponentLocaleProperty(component));
    }

    private CTComponentLocaleProvider(CTObservableProperty<SUPPORTED_LOCALES> locale) {
        this.locale = locale;
    }

    @Override
    public Locale getLocale() {
        return locale.get().getLocale();
    }

    @Override
    public void setLocale(Locale locale) {
        this.locale.set(SUPPORTED_LOCALES.forLocale(locale));
    }

    @Override
    public CTObservableProperty<SUPPORTED_LOCALES> locale() {
        return locale;
    }
}

File: src/main/kotlin/gargoyle/ct/messages/impl/CTComponentLocaleProvider.kt
package gargoyle.ct.messages.impl

import gargoyle.ct.messages.LocaleProvider
import gargoyle.ct.pref.CTPreferences.SupportedLocales
import gargoyle.ct.prop.CTObservableProperty
import gargoyle.ct.ui.prop.CTComponentLocaleProperty
import java.awt.Component
import java.util.Locale

class CTComponentLocaleProvider private constructor(private val locale: CTObservableProperty<SupportedLocales>) :
    LocaleProvider {
    constructor(component: Component) : this(CTComponentLocaleProperty(component))

    override fun getLocale(): Locale {
        return locale.get().locale
    }

    override fun setLocale(locale: Locale) {
        this.locale.set(SupportedLocales.forLocale(locale))
    }

    override fun locale(): CTObservableProperty<SupportedLocales> {
        return locale
    }
}
